### PR TITLE
Provision hypervisor : Morpheus-122

### PR DESCRIPTION
**Pull request by rahulray19**

### CODE REVIEW

File: src/main/groovy/com/morpheusdata/hyperv/HyperVApiService.groovy [hunk 1]
```diff
@@ -21,7 +20,9 @@ class HyperVApiService {
     static defaultRoot = 'C:\\morpheus'
 
     def executeCommand(command, opts) {
-        def output = morpheusContext.executeWindowsCommand(opts.sshHost, opts.sshPort?.toInteger(), opts.sshUsername, opts.sshPassword, command, null, false).blockingGet()
+        // TODO: default sshPort value need to be removed from below code after appliance update
+        def sshPort = opts.sshPort ? opts.sshPort?.toInteger() : 5985
```
Comment (ddeyoung) [inline src/main/groovy/com/morpheusdata/hyperv/HyperVApiService.groovy]: Complete the above todo and validate the correct behavior.

File: src/main/groovy/com/morpheusdata/hyperv/HyperVApiService.groovy [hunk 2]
```diff
@@ -615,7 +616,7 @@ class HyperVApiService {
             def out = wrapExecuteCommand(command, opts)
             log.debug("out: ${out.data}")
             if (out.success) {
-                hasMore = out.data != ''
+                hasMore = out.data != null
```
Comment (ddeyoung) [inline src/main/groovy/com/morpheusdata/hyperv/HyperVApiService.groovy:617]: this should be '' or null

File: src/main/groovy/com/morpheusdata/hyperv/HyperVCloudProvider.groovy [hunk 3]
```diff
@@ -768,21 +826,23 @@ class HyperVCloudProvider implements CloudProvider {
 	}
 
 	private resolveUniqueIdsToVMids(Map opts, node) {
+		log.debug("opts: ${opts}")
 		DataQuery query = new DatasetQuery().withFilters(
-				new DataFilter("zone", opts.zone),
+				new DataFilter("zone.id", opts.zone?.id),
```
Comment (ddeyoung) [inline src/main/groovy/com/morpheusdata/hyperv/HyperVCloudProvider.groovy:829]: this is not going to return the correct results if zone.zone is null. What is the expected behavior if zone ID is 1 vs if zone is null?

File: src/main/groovy/com/morpheusdata/hyperv/HyperVCloudProvider.groovy [hunk 4]
```diff
@@ -479,21 +519,20 @@ class HyperVCloudProvider implements CloudProvider {
...
+					log.debug("hostOnline: ${hostOnline}")
 					if (hostOnline) {
 						allOnline = hostOnline && allOnline
 						anyOnline = hostOnline || anyOnline
 						//check if this host is online
 						def hostResults = loadHypervHost([zone: cloud], hypervisor)
-						if (hostResults.success == true && hostResults.host && hostResults.host['ComputerName']) {
+						if (hostResults.success == true && hostResults.host && hostResults?.host['ComputerName']) {
 							resolveUniqueIdsToVMids([zone: cloud], hypervisor)
-							def results = listVirtualMachines(hypervOpts)
```
Comment (ddeyoung) [inline src/main/groovy/com/morpheusdata/hyperv/HyperVCloudProvider.groovy:493]: listVirtualMachines and related should all be in a VirtualMachineSync service.

**Review by ddeyoung — CHANGES_REQUESTED** [review]

File: src/main/groovy/com/morpheusdata/hyperv/HyperVApiService.groovy (hunk 2 above)
Comment (rahulray19) [reply src/main/groovy/com/morpheusdata/hyperv/HyperVApiService.groovy]: review comment changes added as suggested

File: src/main/groovy/com/morpheusdata/hyperv/HyperVCloudProvider.groovy (hunk 3 above)
Comment (rahulray19) [reply src/main/groovy/com/morpheusdata/hyperv/HyperVCloudProvider.groovy:833]: null check added for zone before query

File: src/main/groovy/com/morpheusdata/hyperv/HyperVCloudProvider.groovy (hunk 4 above)
Comment (rahulray19) [reply src/main/groovy/com/morpheusdata/hyperv/HyperVCloudProvider.groovy:493]: to test the connection, fetching only one virtual machine

File: src/main/groovy/com/morpheusdata/hyperv/HyperVCloudProvider.groovy (hunk 3 above)
Comment (ddeyoung) [reply src/main/groovy/com/morpheusdata/hyperv/HyperVCloudProvider.groovy:833]: This whole method looks like it belongs in the virtual machine sync service, so we'll leave this for now and resolve this in the cache virtual machine story.

File: src/main/groovy/com/morpheusdata/hyperv/HyperVCloudProvider.groovy [hunk 5]
```diff
@@ -794,31 +856,31 @@ class HyperVCloudProvider implements CloudProvider {
...
 					server.uniqueId = match.ID
-					//server.save(flush: true)
 					context.services.computeServer.save(server)
 				}
 			}
 		}
 	}
 
-	def listVirtualMachines(opts) {
+	def listVirtualMachines(opts, testVmConnFlg) {
```
Comment (ddeyoung) [inline src/main/groovy/com/morpheusdata/hyperv/HyperVCloudProvider.groovy:865]: I don't think we need this method, just call the API service directly.

**Review by ddeyoung — CHANGES_REQUESTED** [review]

File: src/main/groovy/com/morpheusdata/hyperv/HyperVApiService.groovy [hunk 6]
```diff
@@ -546,12 +545,12 @@ class HyperVApiService {
         return rtn
     }
 
-    def listVirtualMachines(opts) {
+    def listVirtualMachines(opts, Integer pageSizeParam, testVmConnFlg) {
         def rtn = [success: false, virtualMachines: []]
```
[the remark this answers is not in the document]
Comment (rahulray19) [reply src/main/groovy/com/morpheusdata/hyperv/HyperVApiService.groovy:549]: review comment changes added as suggested

File: src/main/groovy/com/morpheusdata/hyperv/HyperVCloudProvider.groovy (hunk 5 above)
Comment (rahulray19) [reply src/main/groovy/com/morpheusdata/hyperv/HyperVCloudProvider.groovy]: review comment changes added as suggested

**Review by ddeyoung — APPROVED** [review]

**Pull request merged**